### ISSUE TITLE
Add multi-domain support

### ISSUE THREAD

**Issue by ropetin** [eclecticiq/OpenTAXII#191]
(edit by @erwin-eiq) 
This question has been turned into a feature request. Description in [comment](https://github.com/eclecticiq/OpenTAXII/issues/191#issuecomment-995947375)

****
(original question)

Hopefully I'm not missing something obvious in the documentation, but I can't find a definitive answer for this question.  Is it possible to set a single OpenTAXII server with multiple OPENTAXXI_DOMAIN / domain settings?  I need to have the same data available from multiple endpoints on different IPs, ports, and domain names.  I can route the traffic fine with my firewall and HAProxy, but obviously that doesn't help changing the URLs sent back with poll requests, for example.

**Comment by arcsector** [comment]
Hi @ropetin,

Are you deploying on docker or a VM? Regardless, you can set the opentaxii variable to `0.0.0.0` and the server will listen on whatever name it requires. This is even more trivial when you utilize an http proxy in order to listen on a specific name. The web architecture for a distributed deployment would have to be coordinated by yourself

**Comment by ErwinJunge** [comment]
I think @ropetin is talking about the service addresses in the output including the configured domain and wanting to have multiple domain support there.

If so, the current answer is "no, or at least not easily". It's not supported directly in the opentaxxi code. To achieve the goal, there are essentially 2 options:
1) Set up 2 instances of opentaxii with the same db connection parameters, but different domain settings. Then use your reverse proxy to select the correct one. This is achievable now, but obviously a bit wasteful on resources.
2) Build host-header checking into opentaxii to detect the hostname at runtime. This would additionally need a whitelist check against the configured domain(s) for security reasons. This is not achievable with the current release of opentaxii, but would lead to the desired result with only 1 instance of opentaxii running.

@ropetin can you confirm that this is what you meant or clarify the question if it is not? Assuming it is, can you additionally comment on the possible solutions?

**Comment by ropetin** [comment]
> @ropetin can you confirm that this is what you meant or clarify the question if it is not? Assuming it is, can you additionally comment on the possible solutions?

This is exactly what I'm talking about, yes.  Option one is one of the things I tried, but I could not get it to work as expected.  I tried three backends; SQLite3, MariaDB, and PostgreSQL, all with the same result: only one instance could successfully use the data at any time.  I.e. I'd spin up the first instance, get some data in and everything would work like a champ.  I'd then spin up the second instance, connect it to the backend and that would work fine, while the first instance started failing.  Restarting the first instance would get that working again, but the second would now fail.  Rinse and repeat.  I don't have the error specifics because I destroyed that infrastructure, but can spin it back up if it would be helpful.

Option two definitely feels like the best solution and how I think most users would expect it to work.  I don't believe this can be achieved outside of OpenTAXII (say with HAProxy) because it would require rewriting actual payloads.

**Comment by erwin-eiq** [comment]
I agree that option 2 is the best solution. It's not currently possible however. We need code changes that do the following things:

- Accept a list of domains in the config
- Check hostname of request against list, fail if it's not in the list (whitelist approach)
- Use hostname of request when generating response

Let's turn this issue into a feature request for that.

I'm currently busy with adding taxii2 support, which has higher priority than this feature request so even though this is definitely a feature I'm willing to support I can't work on it right now. However, I'll gladly review and potentially accept a PR if you have time sooner than I do.
